File: day06/.ts
type Glove = { hand: 'L' | 'R'; color: string }

function matchGloves(gloves: Glove[]): string[] {
  const pending = {};
  const result = [];

  for (const glove of gloves) {
    const hand = glove.hand;
    const color = glove.color;

    if (!pending[color]) {
      pending[color] = { L: 0, R: 0 };
    }

    if (hand === 'L' && pending[color].R > 0) {
      pending[color].R--;
      result.push(color);
    } else if (hand === 'R' && pending[color].L > 0) {
      pending[color].L--;
      result.push(color);
    } else {
      pending[color][hand]++;
    }
  }

  return result;
}

File: day06/.py
from typing import List, Dict

def match_gloves(gloves: List[Dict[str, str]]) -> List[str]:
  pending = {}
  result = []

  for glove in gloves:
    color = glove["color"]
    hand = glove["hand"]

    if color not in pending:
      pending[color] = {"L": 0, "R": 0}

    if hand == "L" and pending[color]["R"] > 0:
      pending[color]["R"] -= 1
      result.append(color)

    elif hand == "R" and pending[color]["L"] > 0:
      pending[color]["L"] -= 1
      result.append(color)
      
    else:
      pending[color][hand] += 1

  return result